### ISSUE TITLE
ALE `prettier` fixer not processing based on FileType

### ISSUE THREAD

**Issue by chaucerbao** [w0rp/ale#1515]
## Information

**VIM version**

VIM - Vi IMproved 8.0 (2016 Sep 12, compiled Apr 13 2018 10:30:54)
macOS version

Operating System: macOS High Sierra 10.13.4

### :ALEInfo

```
 Current Filetype: 
Available Linters: []
  Enabled Linters: []
 Linter Variables:

 Global Variables:

let g:ale_cache_executable_check_failures = 0
let g:ale_change_sign_column_color = 0
let g:ale_command_wrapper = ''
let g:ale_completion_delay = 100
let g:ale_completion_enabled = 1
let g:ale_completion_max_suggestions = 50
let g:ale_echo_cursor = 1
let g:ale_echo_msg_error_str = 'Error'
let g:ale_echo_msg_format = '%code: %%s'
let g:ale_echo_msg_info_str = 'Info'
let g:ale_echo_msg_warning_str = 'Warning'
let g:ale_enabled = 1
let g:ale_fix_on_save = 0
let g:ale_fixers = {'typescript': ['prettier'], 'markdown': ['prettier'], 'scss': ['prettier'], 'json': ['prettier'], 'javascript': ['prettier'], 'css': ['prettier']}
let g:ale_history_enabled = 1
let g:ale_history_log_output = 1
let g:ale_keep_list_window_open = 0
let g:ale_lint_delay = 200
let g:ale_lint_on_enter = 1
let g:ale_lint_on_filetype_changed = 1
let g:ale_lint_on_save = 1
let g:ale_lint_on_text_changed = 'always'
let g:ale_lint_on_insert_leave = 0
let g:ale_linter_aliases = {}
let g:ale_linters = {}
let g:ale_linters_explicit = 0
let g:ale_list_window_size = 10
let g:ale_list_vertical = 0
let g:ale_loclist_msg_format = '%code: %%s'
let g:ale_max_buffer_history_size = 20
let g:ale_max_signs = -1
let g:ale_maximum_file_size = 0
let g:ale_open_list = 0
let g:ale_pattern_options = {}
let g:ale_pattern_options_enabled = 0
let g:ale_set_balloons = 0
let g:ale_set_highlights = 1
let g:ale_set_loclist = 1
let g:ale_set_quickfix = 0
let g:ale_set_signs = 1
let g:ale_sign_column_always = 0
let g:ale_sign_error = '>>'
let g:ale_sign_info = '--'
let g:ale_sign_offset = 1000000
let g:ale_sign_style_error = '>>'
let g:ale_sign_style_warning = '--'
let g:ale_sign_warning = '--'
let g:ale_statusline_format = ['%d error(s)', '%d warning(s)', 'OK']
let g:ale_type_map = {}
let g:ale_use_global_executables = v:null
let g:ale_warn_about_trailing_blank_lines = 1
let g:ale_warn_about_trailing_whitespace = 1
  Command History:
```

## What went wrong

I have `g:ale_fixers ` set:
```
let g:ale_fixers={
	\'css': ['prettier'],
	\'javascript': ['prettier'],
	\'json': ['prettier'],
	\'markdown': ['prettier'],
	\'scss': ['prettier'],
	\'typescript': ['prettier']
\}
```

It works on corresponding **files** of the respective types. However, if I start with an empty buffer and use `:setfiletype`, it seems that only 'javascript' and 'typescript' **filetypes** will be processed. Nothing happens for the rest.

## Reproducing the bug

Steps for repeating the bug:

1. Start vim
2. `:setfiletype css`
3. Type in some CSS that isn't formatted
    ```css
    body{margin:0;padding:0;}
    ```
4. `:ALEFix`

Nothing happens. Let's try Javascript.

1. Open a new, empty buffer
2. `:setfiletype javascript`
3. Type in some Javascript that isn't formatted
    ```js
    function test(){console.log('Test')}test()
    ```
4. `:ALEFix`

This gets formatted.

**Comment by w0rp** [comment]
Gotcha. ALE did use the `--parser` option at some point, but someone removed the code for that. I'll add it back again.

**Comment by butlerx** [comment]
prettier doesn't require the use of `--parser` anymore as it infers it now.

> Both the babylon and flow parsers support the same set of JavaScript features (including Flow). Prettier automatically infers the parser from the input file path, so you shouldn't have to change this setting.

**Comment by chaucerbao** [comment]
You need `--parser` based on the FileType if there's no filename (ie new buffer).

For example, if you grab compressed JSON output from an API and you want to format it for debugging.

**Comment by w0rp** [comment]
Yep, `--parser` is needed where `prettier` can't figure out what type of file you have, but where Vim knows what type of file it is.

**Comment by chaucerbao** [comment]
@w0rp PR available at #1620. I've only updated the existing tests to account for the `--parser` option. It would be nice to have specific tests against supported filetypes, but my VimFu is elementary level.
